File: 303-range-sum-query-immutable/solution.py
class NumArray:

    def __init__(self, nums: List[int]):
        self.presum = [0] * (len(nums)+1)
        for i in range(1, len(nums)+1):
            self.presum[i] = self.presum[i-1] + nums[i-1]

    def sumRange(self, left: int, right: int) -> int:
        return self.presum[right+1] - self.presum[left]
        

# Your NumArray object will be instantiated and called as such:
# obj = NumArray(nums)
# param_1 = obj.sumRange(left,right)

File: 303-range-sum-query-immutable/solution.cpp
class NumArray {
public:
    NumArray(vector<int>& nums) {
        int n = nums.size();
        preSum = vector<int>(n + 1, 0);
        for (int i = 1; i <= n; i ++) {
            preSum[i] = preSum[i-1] + nums[i-1];
        }
    }
    
    int sumRange(int left, int right) {
        return preSum[right+1] - preSum[left];
    }

private:
    vector<int> preSum;
};

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray* obj = new NumArray(nums);
 * int param_1 = obj->sumRange(left,right);
 */